### ISSUE TITLE
Only verified alarms will trigger panel to be "triggered"?

### ISSUE THREAD

**Issue by arrelid** [mbrrg/pyspcwebgw#9]
Hi there,

Alarm newbie here, so please be gentle...

I've got the spc-webgateway up and running, and almost everything is working like I had assumed it would. Except one thing: if I arm a panel to away mode, and one of its zones trigger a little bit later, the corresponding panel in HA won't get triggered? From what I can tell, I need a second zone to be triggered - which I believe is referred to as a verified alarm - before the panel in HA switches to "triggered" as its state?

Here's a log from a dummy attempt just now. The corresponding alarm panel In HA doesn't change to "triggered" until the second zone (timestamp 1612792885 ) is triggered.

`{"time":"1612792812","raw":"040301000F27FD00","text":"Område 1 Main House Tillkoppla av Användare 9999 Engineer -WEB"},
{"time":"1612792851","raw":"0200040000000000","text":"Larm: Sektion 4 105 Motion Larm"},{"time":"1612792851","raw":"0247010003000000","text":"Larm: FÖRSTA SEKTIONEN 4 - 105 Motion , Område 1 Main House"},
{"time":"1612792885","raw":"02000D0000000000","text":"Larm: Sektion 13 110 Motion Larm"},{"time":"1612792885","raw":"021A010000000000","text":"Larm: Verifierat larm, Område 1 Main House"},{"time":"1612792886","raw":"0200090000000000","text":"Larm: Sektion 9 101 Motion Larm"}]}}}`

Maybe this is by design? That said, I'd love to be able to catch "all" (i.e. pre-verified) alarms too. I can't find any option to configure this in the integration itself, so I would assume that I'd need to change something in the configuration of the alarm panel itself for this to be the case...?

Any pointers would be more than appreciated.

**Comment by crallian** [comment]
Yes, for some reason pyspcwebgw triggers on the sia code 'BV' burglary verified, usually sent after the second sensor is triggered. Note that I think fire alarm etc. never sends BV, just the burglary sensors. It is quite easy to change in a custom integration if you like to.

**Comment by arrelid** [comment]
> Yes, for some reason pyspcwebgw triggers on the sia code 'BV' burglary verified, usually sent after the second sensor is triggered. Note that I think fire alarm etc. never sends BV, just the burglary sensors. It is quite easy to change in a custom integration if you like to.

I was leaning towards doing something like that, but thought I should ask first.

Would need to do some testing on the fire alarms too - doesn't look like the ['FA'](http://doc.vanderbiltindustries.com/docs/Intrusion/SPC/SPCPanel/v_3.8.5/InstallAndConfig_OLH/EN/Content/Topics/SIA_Codes.htm?TocPath=Appendix%7C_____7) or any related (for fire alarms) SIA codes are caught...

**Comment by crallian** [comment]
No, nothing for fire alarms is cought, but you get the status of the sensors. To change in pyspcgw you have to make both spc and pyspcgw custom components, spc just to point to your custom pyspcgw. I can send you or maybe publish my code for this later... maybe in the evening.

**Comment by arrelid** [comment]
Any reason why one wouldn't want these changes upstream (i.e. this repo)?

**Comment by crallian** [comment]
I agree, I would like to have this change too... but haven't done it yet, I decided to try to not have too much difference from the master branch, and avoid braking changes. Instead I use a technical sensor in SPC to flag alarm triggered, this way I can can control the behavior precisely.
Also, @mbrrg, with all respect, seems busy with other things, so I still havn't got any feedback on my pull request from last year... so I decided to not try to do too much local stuf...
If you would like to make your own changes locally, I suggest you copy this folder spc to "custom_components"
https://github.com/crallian/home-assistant/tree/spc_local_custom_component/homeassistant/components/spc
The only change is that it referres to a local pyspcwebgw
And then also copy the folder "pyspcwebgw" to custom coponents, where you can make your changes.
You can use the master branch by mbrrg, or my pull request that solves all the open issues and a few more. The only "breaking change", if I remember correctly, is that alarm set by user... now shows N/A for partset. The reason for this is that the user ID for partset is not part of the response, so the wrong ID was used in some cases.
Note that I have not been runnning the exact code in my repo by now, I just updated with later commits (some minor tidy up) without any conflicts, so it should work... I am migrating to a new server for HassOS at the moment, so I will try myself within a few days... the code before this last updating commit was however run by me and a few others for a year or so...
I think the minor, dirty, way to change this would be to just change line 53 in pyspcwebgw/area.py
        self._verified_alarm = sia_code == 'BV'
to whatever you like.. a nicer way would be to update the attribute to something like "triggered_alarm" instead, requiring a similar change in spc/alarm_control_panel.py and other platforms that might use spcwebgw... so this would require a commit to Home Assistant as well... but it would be nice... and fits better with the STATE_ALARM_TRIGGERED attribute in Hoem Assistant...

**Comment by crallian** [comment]
...and now I am running the code exactly as in the shared repos above, and without testing too much, the custom_component seems to be working as it should.
(Note that I have not done anyting with the 'BV'... but it should be easy... feel free to ask if you need any guidance or want to discuss further)

**Comment by mbrrg** [comment]
Sorry for being "a bit" absent from this. Thanks for your contributions, I will try to have a look tonight or in the next days so that you don't have to create workarounds.

**Comment by mbrrg** [comment]
1) Unfortunately my memory fails me when it comes to the decision to only trigger the alarm state on the HA panel when the alarm is reported as verified. I guess it depends a bit on how trigger happy you want the system to be, I think there are also SPC options to only make the sirens activate after a second activation etc. and also some attribute on individual sensors? A good (better?) way to handle it could perhaps be to remove the BV constraint (in the HA codebase) but to keep the verified alarm attribute. It would then be up to the user to take this into account when writing integrations. Anyway, any change regarding this would be in the HA repo.

2) Suggestions on how to handle a fire alarm are most welcome. Currently you will have to listen to state changes in HA for your fire alarm sensors (as noted above) and act on that but perhaps triggering an alarm on the panel in HA would be more correct, as long as there is some way to distinguish it from a burglar alarm and vice versa. This would require changes both in pyspcwebgw and in the HA spc component.

@crallian how up to date is your pull request here vs additional modifications and cleanup you mention? Would be awesome if you could either update the existing PR or create new ones, preferably on per fix (or at least one fix per commit).

**Comment by arrelid** [comment]
First of all - thanks for making this integration in the first place! It's been absolutely wonderful to be able to close the gap between what's available in the SPC (iOS) app and what you can do with critical notifications in HA <3

As for 1: I would - very selfishly - vote for removing the BV constraint but keeping the verified attribute. This also allows for more flexible HA integrations (but not automations, unless you somehow can expose that attribute to the panel entity), which I'd say is a plus.

As for 2: I will not have opinions on this yet (I'm still too new to say anything smart about anything related to alarms), but the SPC web interface does show you all the sensors that have been triggered since a (single) alarm triggered. In a world where I am dreaming and know nothing about the complexities of such an endeavour, I would say that it would be neat to have something similar, i.e. a list of sensors that are associated with the currently triggered alarm. Right now you can try to be clever and ask all sensors that _you_ know belong to a certain alarm panel to figure out if they have changed state since the alarm triggered, but that can give you false positives (i.e. alarm set to HOME or NIGHT, and it's just fine that you trigger motion sensors on the inside POST an alarm triggering due to a door being opened).

**Comment by crallian** [comment]
@mbrrg 
1. I am not really sure what you mean?
To add an attribute "triggered_alarm" or maybe "burglary_alarm" in pyspcwebgw, and also keep the verified_alarm attribute here. This would enable the possibility to change the reference in HA spc to the new attribute, or add it in HA spc as a config option? That sounds like a good idea!
(I think that if we add more area-alarm-attributes we have to add code to reset/restore the alarm, because currently any area-supported SIA code except BV (today only set/unset codes) will reset BV.)

2. Maybe we can also add a fire-alarm attribute in the same way, when we are at it, to enable the possibility for HA spc? :)

My pull-request is up to date. I have learnt that it was a bad idea to put all changes in one PR. :(
I think however the diffrent changes are quite independent, I will comment under the pull request. If that is not enough to make it clear, I guess I have to re-do everything, in seperate commits?

**Comment by arrelid** [comment]
Sidenote: my robot vacuum managed to trip one of the alarm's sensors in away mode the other day - a gentle reminder that it would be amazing to have a way to distinguish between unverified and verified alarms :)

**Comment by arrelid** [comment]
I'm not sure if you (@crallian and @mbrrg) have continued discussing this off-thread, but wanted to ask at least.

If it looks like this won't be fixed in this repo, I'd love to try the workaround (forked repo + any dependencies) so that my HA instance has a better idea of the state of the alarm panels.

Thanks <3

**Comment by crallian** [comment]
@arrelid Unfortunatelly, I haven’t heard anything from @mbrrg since his last comment here i february.

For your request to see which zones are involved in the active larm, I think that functionallity is actually already implemented in pyspcwebgw, but not in HA.
In pyspcwebgw, every zone has a ZoneType, a ZoneInput (open/close), but also a ZoneStatus that is not implemented in HA. (see files zone.py and const.py)
In HA each zone is just a binary sensor ZoneInput(open/close) of type ZoneType, so we have to think of the best way to implement this, but this should be discussed in the HA integration I think.

For a workaround regarding your issue, you could also do as I do... I use two virtual technical zones i SPC (available from fw 3.11), one for Burglary alarm, and one for Fire Alarm... If it is on, there is at least one sensor if the type giving an alarm... I guess an alternative could be to just create a ”group” in HA, with all sensors of a certain type...

**Comment by mbrrg** [comment]
I have now finally had some more time to investigate this further. The current implementation of pyspcwebgw is pretty dumb (by design), SPC alarm events (SIA: BA) are sent by SPC/webgw for zones only. In the HA world, alarm events are triggered on the panel device. So I agree with you @crallian that this should be solved in the HA integration to keep this close to the spcwebgw representation. In other words, the HA integration should be modified so that alarm status on any on the zones would trigger the alarm for its parent panel.

I have one PR active (actually it was closed because I was too slow) in the HA repo that has quite a lot of modifications to the integration, but once I get that in this could be a very good next thing to look at.

**Comment by arrelid** [comment]
Most excellent news! Let me know if/when there's something you want me to try and I'll happily help out. Otherwise I'll just stay tuned, eagerly awaiting 😄

**Comment by arrelid** [comment]
Summer update: the original problem (as described by this issue) has happened twice this summer - both times while I was away on vacation...  😂

I haven't made any contributions to HA (or any of its components/dependencies) so far, but I might give this a go unless either of you have made progress locally but didn't submit / check in your work yet?

**Comment by crallian** [comment]
> Summer update: the original problem (as described by this issue) has happened twice this summer - both times while I was away on vacation... 😂
> 
> I haven't made any contributions to HA (or any of its components/dependencies) so far, but I might give this a go unless either of you have made progress locally but didn't submit / check in your work yet?

Nope, I have not looked into this any further yet,  but the issue just got real again since I needed my virtual technical zones (max 4 4320) for other stuff…

A quick workaround would be to just locally replace BV with BA, I just did this myself… tell me if you want to copy my config.
